### PR TITLE
dev: MockProver fixes and improvements

### PR DESCRIPTION
**Pull request by str4d**

### CODE REVIEW

File: src/dev.rs [hunk 1]
```diff
@@ -43,16 +44,21 @@ pub enum VerifyFailure {
...
+    /// A constraint was not satisfied for a particular row.
+    Constraint {
+        /// The index of the gate containing the unsatisfied constraint. These indices are
+        /// assigned in the order in which `ConstraintSystem::create_gate` is called
+        /// during `Circuit::configure`.
         gate_index: usize,
-        /// The name of the gate that is not satisfied. These are specified by the gate
-        /// creator (such as a chip implementation), and may not be unique.
+        /// The name of the gate containing the unsatisfied constraint. This is specified
+        /// by the gate creator (such as a chip implementation), and may not be unique.
```
Comment (daira) [inline src/dev.rs:54]:
```suggestion
        /// by the gate creator (such as a chip implementation), and is not enforced to
        /// be unique.
```

File: src/dev.rs [hunk 2]
```diff
@@ -379,7 +431,8 @@ impl<F: FieldExt> MockProver<F> {
                                     Any::Fixed => {
```
Comment (daira) [inline src/dev.rs:431]: (applies to lines 418-420 above)

I didn't know that selectors could be on a different row to the gate's logical row! That seems confusing. Is it necessary at all, or should we enforce that selectors have a row offset of 0?

Comment (str4d) [reply src/dev.rs:431]: Now that the `Selector` type consistently means a positive boolean selector, _maybe_ rotations are unnecessary for them? I'm fine with restricting this if desired.

Comment (daira) [reply src/dev.rs:432]: It is technically possible that there could be optimizations where you share selectors that are on different rows between gates, but the more I think about how confusing that would be, the more I think it should be prohibited. It also potentially interferes with visualizations where you click on a row and see the gates enabled for that row.

(If you *really* want to do those optimizations, you can still do them using fixed columns.)

Comment (str4d) [reply src/dev.rs:432]: I'll open a separate PR for this.

Comment (str4d) [reply src/dev.rs:432]: https://github.com/zcash/halo2/pull/287

**Review by ebfull — APPROVED** [review]

**Review by daira — COMMENTED** [review]
Post-hoc ACK 483c872cbeb8956f87ec2cb33f71a4a3622b919c

**Pull request merged**